README.md
# waqiapi

*WAQIAPI* is a Flutter package that can be used to access the World Air Quality Index (WAQI) [public JSON API](https://aqicn.org/json-api/doc/). To access the API you will need to register and obtain a token, more details can be found [here](https://aqicn.org/data-platform/token/#/).


## Using the API
Just add the package to your project then import into you code, see below.

```
import 'package:waqiapi/waqiapi.dart' as WaqiApi
```

You will then have access to the API along with a helper GpsCoord class. To use the API just define the following (using a valid token):

```
WaqiApi.Api api = WaqiApi.Api(token: "{Put a Valid Token In Here or It Will Error}");
```

You can define sets of Gps coords using the helper class: 

```
WaqiApi.GpsCoord gpsa = WaqiApi.GpsCoord(lat: 39.379436, long: 116.091230);
WaqiApi.GpsCoord gpsb = WaqiApi.GpsCoord(lat: 40.235643,long: 116.784382);
```

These can then be used to pass into the API. The following are valid calls available in the API (all are async) and all return a JSON object:

```
var feedResult = await api.cityFeed("London");
var ipResult = await api.locationFeedIPAddress();
var gpsResult = await api.locationFeedLatLng(gpsa);
var queryResult = await api.mapQuery(gpsa, gpsb);
var nameResult = await api.searchByName("Bloomsbury");
```


See the example project for more details.





## Getting Started

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [0.1.1] - 07 Sep 2019

* Tidy up pubspec and format dart source.

## [0.1.0] - 07 Sep 2019

* Initial release.


lib/src/api.dart
// Copyright (c) 2019, Steve Rogers. All rights reserved. Use of this source code
// is governed by an Apache License 2.0 that can be found in the LICENSE file.
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

/// Class representing a GPS coordinate using [lat] and [long]
class GpsCoord {
  final double lat;
  final double long;
  GpsCoord({@required this.lat, @required this.long});
}

/// API to access the WAQI API -  A valid [token] is required to use the API.
/// See https://aqicn.org/api/ and https://aqicn.org/json-api/doc/ for more details
class Api {
  final String token;
  final String _endpoint = "https://api.waqi.info";

  Api({@required this.token});

  /// Internal call to query our API
  _callAPI(String url) async {
    var result = await http.get(Uri.encodeFull(url));
    if (result.statusCode == 200) {
      return jsonDecode(result.body);
    } else {
      throw Exception(
          "WAQIAPI API Error calling $url. Status code: ${result.statusCode}");
    }
  }

  /// Get the real-time Air Quality index for a given station in a [city]
  cityFeed(String city) async {
    String apiUrl = "$_endpoint/feed/$city/?token=$token";
    return _callAPI(apiUrl);
  }

  /// Search for the nearest station from a given latitude/longitude defined by [gps]
  locationFeedLatLng(GpsCoord gps) async {
    String apiUrl = "$_endpoint/feed/geo:${gps.lat};${gps.long}/?token=$token";
    return _callAPI(apiUrl);
  }

  /// Get the nearest station close to the user location, based on the IP adress information.
  locationFeedIPAddress() async {
    String apiUrl = "$_endpoint/feed/here/?token=$token";
    return _callAPI(apiUrl);
  }

  /// Get all the stations within a given lat/lng bounds defined by [coords1]  and [coords2]
  mapQuery(GpsCoord coords1, GpsCoord coords2) async {
    String apiUrl =
        "$_endpoint/map/bounds/?latlng=${coords1.lat},${coords1.long},${coords2.lat},${coords2.long}&token=$token";
    return _callAPI(apiUrl);
  }

  /// search stations by [name]
  searchByName(String name) async {
    String apiUrl = "$_endpoint/search/?keyword=$name&token=$token";
    return _callAPI(apiUrl);
  }
}


example/lib/main.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:waqiapi/waqiapi.dart' as WaqiApi;

/// This is DEMO app put togther to show the use of the Waqi API only
void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter WAQI API Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter WAQI API Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  // define the API access here - provide a valid TOKEN!
  WaqiApi.Api api = WaqiApi.Api(token: "{Put a Valid Token In Here or It Will Error}");

  // Some test GPS Coords from the WAQI API docs
  WaqiApi.GpsCoord gpsa = WaqiApi.GpsCoord(lat: 39.379436, long: 116.091230);
  WaqiApi.GpsCoord gpsb = WaqiApi.GpsCoord(lat: 40.235643,long: 116.784382);

  // Var for holding result of API call
  String resultData;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'Result of API Call:',
            ),
            Text(
              '$resultData'
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () async {

          // The available API's
          //var result = await api.cityFeed("London");
          //var result = await api.locationFeedIPAddress();
          //var result = await api.locationFeedLatLng(gpsa);
          //var result = await api.mapQuery(gpsa, gpsb);
          var result = await api.searchByName("Bloomsbury");

          // display any result
          print(result);
          setState(() {
            resultData = jsonEncode(result);
          });

        },
        tooltip: 'CallAPI',
        child: Icon(Icons.refresh),
      ), 
    );
  }
}
